orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java

```diff
@@ -0,0 +1,559 @@
+				+ tupleContext.getClass() );
+		return new Tuple( new OrientDBTupleSnapshot( tupleContext.getAllAssociatedEntityKeyMetadata(), tupleContext.getAllRoles(), entityKeyMetadata ) );
+	}
+
+	@Override
+	public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext) throws TupleAlreadyExistsException {
+		log.info( "insertOrUpdateTuple:EntityKey:" + key + "; tupleContext" + tupleContext + "; tuple:" + tuple );
+		Connection connection = provider.getConnection();
+
+		StringBuilder queryBuffer = new StringBuilder();
+		String dbKeyName = key.getColumnNames()[0];
+		Object dbKeyValue = key.getColumnValues()[0];
+
+		for ( int i = 0; i < key.getColumnNames().length; i++ ) {
+			String columnName = key.getColumnNames()[i];
+			Object columnValue = key.getColumnValues()[i];
+			log.info( "EntityKey: columnName: " + columnName + ";columnValue: " + columnValue + " (class:" + columnValue.getClass().getName() + ");" );
+		}
+		try {
+			boolean existsPrimaryKey = EntityKeyUtil.existsPrimaryKeyInDB( provider.getConnection(), key );
+			log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " exists in database ? " + existsPrimaryKey );
+
+			if ( existsPrimaryKey ) {
+				// it is update
+				queryBuffer.append( "update  " ).append( key.getTable() ).append( "  set " );
+
+				for ( String columnName : tuple.getColumnNames() ) {
+					if ( OrientDBConstant.SYSTEM_FIELDS.contains( columnName ) || columnName.equals( dbKeyName ) ) {
+						continue;
+					}
+					// @TODO correct type
+					queryBuffer.append( " " ).append( columnName ).append( "=" );
+					EntityKeyUtil.setFieldValue( queryBuffer, tuple.get( columnName ) );
+					queryBuffer.append( "," );
+				}
+				queryBuffer.setLength( queryBuffer.length() - 1 );
+				queryBuffer.append( " WHERE " ).append( dbKeyName ).append( "=" );
+				EntityKeyUtil.setFieldValue( queryBuffer, dbKeyValue );
+			}
+			else {
+				// it is insert with business key which set already
+				log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " is new! Insert new record!" );
+				queryBuffer.append( "insert into " ).append( key.getTable() ).append( "  set " );
+				for ( String columnName : tuple.getColumnNames() ) {
+					if ( OrientDBConstant.SYSTEM_FIELDS.contains( columnName ) ) {
+						continue;
+					}
+					// @TODO correct type
+					queryBuffer.append( " " ).append( columnName ).append( "=" );
+					EntityKeyUtil.setFieldValue( queryBuffer, tuple.get( columnName ) );
+					queryBuffer.append( "," );
+				}
+				queryBuffer.append( dbKeyName ).append( " = " );
+				EntityKeyUtil.setFieldValue( queryBuffer, dbKeyValue );
+			}
+
+			log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " (" + dbKeyValue + ").  query:" + queryBuffer.toString() );
+			PreparedStatement pstmt = connection.prepareStatement( queryBuffer.toString() );
+			log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " (" + dbKeyValue + "). inserted or updated: " + pstmt.executeUpdate() );
```

Comment:
Shouldn't the query already be in OrientDBEntityQueries?All the EnittyKeyMetadatas are generated at startup, you can create all the queries you need for CRUD operations and cache them.